### DISCUSSION TITLE
Access refs in child component from parent

### DISCUSSION THREAD

**Discussion by roncallyt** [alpinejs/alpine#3000] [1. Help] [answered]
Hi, I would like to access the refs in child component from parent, is this possible?

```html
<div x-data="{
    init() {
        console.log(this.$refs.select);  // Output: undefined
    }
}">                                      // Parent Component
    <div x-data>                         // Child Component
        <select x-ref="select"></select> // Select
    </div>
</div>
```

**Comment by edimamatthew** [comment]
@roncallyt Did you fix this?

**Comment by acentrist** [comment]
@roncallyt  It seems like the same problem with https://github.com/alpinejs/alpine/discussions/1692

**Comment by tlaverdure** [comment]
Here's one way to accomplish this:

```html
<div x-data="{
    child: {},
    init() {
        this.$nextTick(() => console.log(this.child.$refs.select));
    }
}"> 
    <div x-data x-init="child = $data">
        <select x-ref="select"></select>
    </div>
</div>
```

**Note:** `this.child` will not be accessible until the child runs `x-init` and assigns itself to the parent scope, hence the `$nextTick`.

**Comment by marshnmedia** [comment]
@tlaverdure , that is an interesting solution.  However, the solution does not lend itself to building reusable components.  In a reusable component, the child element should have no idea what variables the parent container might or might not have.

Is there really no way to reference x-ref elements in child components?  That seems like something we should have.

**Reply by SimoTod** [reply]
`$root.querySelector('[x-ref="..."]')`

**Comment by ekwoka** [comment, marked as answer]
refs are registered on the nearest component, they don't bubble up higher than that.

Realistically, the parent should be agnostic to the child implementation as well, so having the refs go up higher doesn't make too much sense.

The two should be interacting through boundary APIs, like x-modelable.

Really, refs are an extremely unimportant feature and have almost no use beyond passing an element into third party scripts that need an element passed in. It's likely using a ref for whatever you are doing is a bad way to do it anyway.
